### DISCUSSION TITLE
Feature request: optionally sow attention weight in `dot_product_attention`

### DISCUSSION THREAD

**Discussion by JyChang012** [google/flax#2869] [Ideas] [answered]
It seems that there is no good way to retrieve attention weight using `MultiHeadDotProductAttention`. Probably we can add an optional `mdl` parameter to  `dot_product_attention` and `dot_product_attention_weights`, and `sow` the weight inside. This should have no effect when `intermediates` is not mutable. Here is my [implementation](https://github.com/JyChang012/espnet/blob/jax/asr_train/espnex/models/transformer/attention.py). I can create a PR if you like.

Update:
Implemented in #3529

**Comment by JyChang012** [comment]
Any ideas on how to extract attention weight other than rewriting the attention module? @cgarciae

**Comment by cgarciae** [comment]
Hey @JyChang012, sorry for the delay. I see that Tensorflow's [MultiHeadAttention](https://www.tensorflow.org/api_docs/python/tf/keras/layers/MultiHeadAttention) has a  `return_attention_scores` flag which would solve your use case, however, Flax doesn't have an equivalent feature even though it would be easy to add.

In the mean time, we do encourage users forking our Modules for their own needs, we try to maintain them readable for this purpose.

@marcvanzee WDYT? Should we consider adding this flag?

**Comment by JyChang012** [comment, marked as answer]
This implementation is applied in PR #3529
